crates/subspace-farmer/src/dsn.rs

```diff
@@ -137,10 +159,42 @@ impl DSNSync for subspace_networking::Node {
             .await
             .map_err(GetPiecesByRangeError::CannotGetClosestPeers)?;
 
-        // select first peer for the piece-by-range protocol
-        let peer_id = *peers
-            .first()
-            .ok_or(GetPiecesByRangeError::NoClosestPiecesFound)?;
+        // select appropriate peer for the piece-by-range protocol
+        if peers.is_empty() {
+            return Err(GetPiecesByRangeError::NoClosestPiecesFound);
+        }
+
+        let mut peer_id: Option<PeerId> = None;
+        for peer_id_candidate in peers {
+            let result = self
+                .send_generic_request(peer_id_candidate, PeerInfoRequest)
+                .await;
```

Comment:
Maybe it would be better to cache those requests and do them once for each peer?

Reply:
The peer status could change and I'm not sure it's 'cachable` in principle.

Reply:
I think we can use some struct which would remove entries after some time. Like this one:
https://docs.rs/ttlhashmap/0.1.0/ttlhashmap/struct.TtlHashMap.html

Reply:
It will not be necessary in the future with push model, fine as is for now